### PR TITLE
rtio: Add user mode, tests, and helpers

### PR DESCRIPTION
**Pull request by teburd**
Adds three syscalls, rtio_sqe_copy_in, rtio_cqe_copy_out, and rtio_submit (akin to io_uring_enter). The only way currently to get submissions in and completions out from a user mode thread is the syscalls through copying. This resolves potential TOCTOU issues.

For a quick comparison, io_uring does in fact copy the submission queue events from user space. The shared ring is there to enable a kernel mode thread to poll the queue entirely avoiding syscalls.

RTIO could still have shared ring queues just like io_uring but would require an executor that is a k_object and upon submission takes the RTIO context, copies SQEs in, copies CQEs out, and internally manages its own state.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

### CODE REVIEW

**Comment by teburd** [comment]
Thank you for all the help in understanding how I might go about doing this @dcpleung

File: include/zephyr/rtio/rtio.h [hunk 1]
```diff
@@ -401,6 +401,45 @@ static inline void rtio_sqe_prep_write(struct rtio_sqe *sqe,
 #define RTIO_IODEV_SQ_DEFINE(name, len) \
 	static RTIO_SPSC_DEFINE(name, struct rtio_iodev_sqe, len)
 
+#define Z_RTIO_DEFINE(name, struct_attrs, exec, sq_sz, cq_sz)	\
+	IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM, (K_SEM_DEFINE(_submit_sem_##name, 0, K_SEM_MAX_LIMIT))) \
+	IF_ENABLED(CONFIG_RTIO_CONSUME_SEM, (K_SEM_DEFINE(_consume_sem_##name, 0, 1)))             \
+	struct_attrs RTIO_SQ_DEFINE(_sq_##name, sq_sz);						   \
+	struct_attrs RTIO_CQ_DEFINE(_cq_##name, cq_sz);						   \
```
[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/rtio/rtio.h:408]: It's not clear but the SQ/CQ DEFINES do that already, probably sensible to move that out of the macro though to make it clearer.

[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/rtio/rtio.h]: Moved the static up

File: subsys/rtio/rtio_handlers.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#ifdef CONFIG_RTIO_SUBMIT_SEM
+	Z_OOPS(Z_SYSCALL_OBJ(r->submit_sem, K_OBJ_SEM));
+#endif
+
+#ifdef CONFIG_RTIO_CONSUME_SEM
+	Z_OOPS(Z_SYSCALL_OBJ(r->consume_sem, K_OBJ_SEM));
+#endif
+
+	/* Validate r is in accessible memory */
+	Z_OOPS(Z_SYSCALL_MEMORY(r, sizeof(struct rtio), true));
```
[the remark this answers is not in the document]
Comment (teburd) [reply subsys/rtio/rtio_handlers.c]: Makes sense, done

File: include/zephyr/rtio/rtio.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcpleung) [reply include/zephyr/rtio/rtio.h]: How about the semaphores? They can be declared static also.

[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/rtio/rtio.h]: doh, yes thanks

**Comment by teburd** [comment]
Some more work needs to go into avoiding TOCTOU issues. Until that's sorted, it's marked as In Progress.

**Comment by teburd** [comment]
I think I've resolved *most* of issues here, the remaining one is the iodev pointer in each SQE isn't being verified and could point at absolutely anything. From a user mode perspective I think this is an issue as it means a kernel address could be given and exploited.

To solve this, probably requires making struct rtio_iodev a k_object as well and validating when copying that it is indeed a valid k_object and that, interestingly, the usermode thread has access to it which is neat.

Some OPs do not need an iodev and I need to fix that as well. There's some general OPs in a batch for timeouts/deadlines/nops/flushes/etc that should really have iodev be NULL and the executor deal with it.

**Comment by teburd** [comment]
rtio_iodev's are now k_objects and have some slightly changed requirements from the original. struct rtio_iodev * is now expected to be const everywhere.

Some cleanup and extra testing needs to happen to validate all the new paths but I'm quite happy with this.

**Review by dcpleung — DISMISSED** [review]

File: subsys/rtio/rtio_handlers.c [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+static inline int z_vrfy_rtio_sqe_copy_in(struct rtio *r,
+					  const struct rtio_sqe *sqes,
+					  size_t sqe_count)
+{
+	printk("checking rtio object permissions\n");
+	Z_OOPS(Z_SYSCALL_OBJ(r, K_OBJ_RTIO));
+
+	printk("checking sqe memory permissions\n");
```
Comment (gmarull) [inline subsys/rtio/rtio_handlers.c:49]: printk leftover?

File: tests/subsys/rtio/rtio_api/CMakeLists.txt [hunk 4]
```diff
@@ -5,7 +5,7 @@ cmake_minimum_required(VERSION 3.20.0)
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 project(rtio_api_test)
 
-FILE(GLOB app_sources src/*.c)
+FILE(GLOB app_sources src/main.c)
```
Comment (gmarull) [inline tests/subsys/rtio/rtio_api/CMakeLists.txt:8]: glob not needed anymore?

File: tests/subsys/rtio/rtio_api/src/main.c [hunk 5]
```diff
@@ -14,6 +14,8 @@
 #include <zephyr/rtio/rtio_executor_concurrent.h>
 
 #include "rtio_iodev_test.h"
+#include "zephyr/app_memory/mem_domain.h"
+#include "zephyr/sys/kobject.h"
```
Comment (gmarull) [inline tests/subsys/rtio/rtio_api/src/main.c:18]: `<>`

File: tests/subsys/rtio/rtio_api/src/rtio_iodev_test.h [hunk 6]
```diff
@@ -35,6 +27,66 @@ struct rtio_iodev_test {
...
+	struct rtio_iodev_test_data *data = CONTAINER_OF(tm, struct rtio_iodev_test_data, timer);
+
+	struct rtio *r = data->r;
+	const struct rtio_sqe *sqe = data->sqe;
+
+	data->r = NULL;
+	data->sqe = NULL;
+
+	/* Complete the request with Ok and a result */
+	printk("sqe ok callback\n");
```
Comment (gmarull) [inline tests/subsys/rtio/rtio_api/src/rtio_iodev_test.h:42]: another printk, if test needs to print use TC_PRINT

File: tests/subsys/rtio/rtio_api/CMakeLists.txt (hunk 4 above)
Comment (teburd) [reply tests/subsys/rtio/rtio_api/CMakeLists.txt:8]: Ah yes, thanks, I was trying to get rid of the glob

**Review by MaureenHelm — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**